File: apps/api/predict/harville.py
"""
Harville and Stern formulas for place and show probabilities.
Based solely on win probabilities (ticket-only safe).
"""
from typing import List, Dict


def harville_place_show(p_win: List[float], use_stern: bool = True) -> List[Dict[str, float]]:
    """
    Compute place and show probabilities using Harville formulas.
    
    Harville formulas:
    - P(place_i) = Σ_{j≠i} [p_i * p_j / (1 - p_i)]
    - P(show_i) = Σ_{j≠i,k≠i,k≠j} [p_i * p_j * p_k / ((1-p_i)(1-p_i-p_j))]
    
    Stern adjustment: Mild flattening for dense favorites (optional).
    
    Args:
        p_win: List of win probabilities (must sum to ~1.0)
        use_stern: Apply Stern adjustment (default True)
    
    Returns:
        List of {p_win, p_place, p_show} dicts
    
    Examples:
    >>> probs = [0.40, 0.30, 0.20, 0.10]
    >>> results = harville_place_show(probs)
    >>> results[0]['p_place'] > results[0]['p_win']  # Place > Win
    True
    """
    n = len(p_win)
    if n < 2:
        # Edge case: only 1 horse
        return [{"p_win": 1.0, "p_place": 1.0, "p_show": 1.0}] if n == 1 else []
    
    # Stern adjustment (optional mild flattening)
    if use_stern:
        # Stern factor: slightly reduce extreme probabilities
        # Formula: p' = p^0.95 (gentle exponent)
        p_adjusted = [p ** 0.95 for p in p_win]
        # Renormalize
        total = sum(p_adjusted)
        if total > 0:
            p_win = [p / total for p in p_adjusted]
    
    results = []
    
    for i in range(n):
        p_i = p_win[i]
        
        # Place probability
        p_place = 0.0
        for j in range(n):
            if j != i:
                denom = 1.0 - p_i
                if denom < 1e-9:
                    denom = 1e-9  # Numerical stability
                p_place += (p_i * p_win[j]) / denom
        
        # Show probability
        p_show = 0.0
        for j in range(n):
            if j == i:
                continue
            for k in range(n):
                if k == i or k == j:
                    continue
                
                denom1 = 1.0 - p_i
                denom2 = 1.0 - p_i - p_win[j]
                
                # Numerical stability
                if denom1 < 1e-9:
                    denom1 = 1e-9
                if denom2 < 1e-9:
                    denom2 = 1e-9
                
                p_show += (p_i * p_win[j] * p_win[k]) / (denom1 * denom2)
        
        # Clamp to valid probability range
        p_place = max(0.0, min(1.0, p_place))
        p_show = max(0.0, min(1.0, p_show))
        
        results.append({
            "p_win": p_i,
            "p_place": p_place,
            "p_show": p_show
        })
    
    return results


def compute_exacta_prob(p_win: List[float], i: int, j: int) -> float:
    """
    Probability of exacta (i first, j second).
    P(i-j) = p_i * p_j / (1 - p_i)
    """
    if i == j or i >= len(p_win) or j >= len(p_win):
        return 0.0
    
    p_i = p_win[i]
    p_j = p_win[j]
    denom = 1.0 - p_i
    
    if denom < 1e-9:
        return 0.0
    
    return (p_i * p_j) / denom

File: lib/harville.js
// lib/harville.js
// ADDITIVE: Harville formulas for place/show probabilities from win probabilities

/**
 * Compute place and show probabilities using Harville formulas.
 * 
 * Harville formulas:
 * - P(place_i) = Σ_{j≠i} [p_i * p_j / (1 - p_i)]
 * - P(show_i) = Σ_{j≠i,k≠i,k≠j} [p_i * p_j * p_k / ((1-p_i)(1-p_i-p_j))]
 * 
 * Args:
 *   winProbs: Array of win probabilities (must sum to ~1.0)
 *   useStern: Apply Stern adjustment (default true)
 * 
 * Returns:
 *   { placeProbs: number[], showProbs: number[] }
 */
export function harvilleFromWinProbs(winProbs, useStern = true) {
  const eps = 1e-9;
  const n = winProbs.length;
  
  if (n < 2) {
    // Edge case: 1 or 0 horses
    if (n === 1) {
      return { placeProbs: [1.0], showProbs: [1.0], winProbs: [1.0] };
    }
    return { placeProbs: [], showProbs: [], winProbs: [] };
  }
  
  // Clamp win probs to [eps, 1-eps] for numerical stability
  let probs = winProbs.map(p => Math.max(eps, Math.min(1 - eps, p || 0)));
  
  // Store original normalized win probs (before Stern adjustment)
  const originalWinProbs = probs.slice();
  
  // Optional Stern adjustment (mild flattening) - only for place/show calculations
  if (useStern) {
    // Stern factor: p' = p^0.95 (gentle exponent)
    probs = probs.map(p => Math.pow(p, 0.95));
    // Renormalize
    const total = probs.reduce((a, b) => a + b, 0);
    if (total > eps) {
      probs = probs.map(p => p / total);
    }
  }
  
  // Compute place probabilities
  // Harville place formula: P(place_i) = P(i finishes 1st) + P(i finishes 2nd)
  // Where:
  // - P(i finishes 1st) = p_i
  // - P(i finishes 2nd) = Σ_{j≠i} [p_j * p_i / (1 - p_j)]
  const placeProbs = [];
  for (let i = 0; i < n; i++) {
    const p_i = probs[i];
    let p_place = 0.0;
    
    // P(i finishes 1st) = p_i
    p_place += p_i;
    
    // P(i finishes 2nd) = Σ_{j≠i} [p_j * p_i / (1 - p_j)]
    for (let j = 0; j < n; j++) {
      if (j !== i) {
        const denom = 1.0 - probs[j];
        if (denom > eps) {
          p_place += (probs[j] * p_i) / denom;
        }
      }
    }
    
    // Clamp to valid range
    placeProbs.push(Math.max(0.0, Math.min(1.0, p_place)));
  }
  
  // Compute show probabilities
  // Harville show formula: P(show_i) = P(i finishes 1st) + P(i finishes 2nd) + P(i finishes 3rd)
  // Where:
  // - P(i finishes 1st) = p_i
  // - P(i finishes 2nd) = Σ_{j≠i} [p_j * p_i / (1 - p_j)]
  // - P(i finishes 3rd) = Σ_{j≠i,k≠i,k≠j} [p_j * p_k * p_i / ((1-p_j)(1-p_j-p_k))]
  const showProbs = [];
  for (let i = 0; i < n; i++) {
    const p_i = probs[i];
    let p_show = 0.0;
    
    // P(i finishes 1st) = p_i
    p_show += p_i;
    
    // P(i finishes 2nd) = Σ_{j≠i} [p_j * p_i / (1 - p_j)]
    for (let j = 0; j < n; j++) {
      if (j === i) continue;
      const denom = 1.0 - probs[j];
      if (denom > eps) {
        p_show += (probs[j] * p_i) / denom;
      }
    }
    
    // P(i finishes 3rd) = Σ_{j≠i,k≠i,k≠j} [p_j * p_k * p_i / ((1-p_j)(1-p_j-p_k))]
    for (let j = 0; j < n; j++) {
      if (j === i) continue;
      for (let k = 0; k < n; k++) {
        if (k === i || k === j) continue;
        const denom1 = 1.0 - probs[j];
        const denom2 = 1.0 - probs[j] - probs[k];
        if (denom1 > eps && denom2 > eps) {
          p_show += (probs[j] * probs[k] * p_i) / (denom1 * denom2);
        }
      }
    }
    
    // Clamp to valid range
    showProbs.push(Math.max(0.0, Math.min(1.0, p_show)));
  }
  
  return { 
    winProbs: originalWinProbs,  // Return original normalized win probs (not Stern-adjusted)
    placeProbs, 
    showProbs 
  };
}

